### ISSUE TITLE
About CapsNet: squash function improve

### ISSUE THREAD

**Issue by PaulGitt** [ageron/handson-ml#173]
Hello, @ageron 
     First, thanks your excellent work about CapsNet it is very helpful.
     I have a suggestion about squash function (not very big deal),
     The squash function can change to this below:

```python
def squash(s, axis=-1, name=None):
    with tf.name_scope(name, default_name="squash"):
        squared_norm = tf.reduce_sum(tf.square(s), axis=axis, keep_dims=True)
        safe_norm = tf.sqrt(squared_norm)
        squash_factor = safe_norm / (1. + squared_norm)
        unit_vector = s
        return squash_factor * unit_vector
```

Then you no need to add epsilon, and it looks more clean.
It didn't conflict in mathematics to me and I also test it in CapsNet, 
it work fine, the result almost as same as the original one.
If I make a mistake, please let me know, thank you.
Thanks again your excellent work.

**Comment by ageron** [comment]
Thanks for your kind words @PaulGitt , I'm glad you find my work helpful.

Unfortunately, this implementation will fail if at any point it encounters a zero vector.  Indeed, the square root of x has no gradient when x=0 (the problematic line is `safe_norm = tf.sqrt(squared_norm)`).

Here's a short code example that demonstrates the problem:

```python
>>> import tensorflow as tf
>>> def squash(s, axis=-1, name=None):
...     with tf.name_scope(name, default_name="squash"):
...         squared_norm = tf.reduce_sum(tf.square(s), axis=axis, keep_dims=True)
...         safe_norm = tf.sqrt(squared_norm)
...         squash_factor = safe_norm / (1. + squared_norm)
...         unit_vector = s
...         return squash_factor * unit_vector
...
>>> s = tf.placeholder(shape=[None], dtype=tf.float32)
>>> r = squash(s)
>>> [g] = tf.gradients(r, s)
>>> with tf.Session() as sess:
...     print(r.eval(feed_dict={s: [2., 3.]}))
...     print(r.eval(feed_dict={s: [0., 0.]}))
...     print(g.eval(feed_dict={s: [2., 3.]}))
...     print(g.eval(feed_dict={s: [0., 0.]}))
...
[ 0.51507872  0.77261806]
[ 0.  0.]
[ 0.08773321  0.00283012]
[ nan  nan]
```

As you can see, your squash function works well, but it returns `nan` when you try to compute the gradient with a zero vector.  It is perfectly possible that training will never encounter a zero vector, in which case it will work fine, but if it encounters a zero vector at any point, it will explode (the variables will become `nan` and then will remain `nan`).

I hope this helps,
Aurélien

**Comment by PaulGitt** [comment]
Oh, I made a big mistake. You are right, it needs a epsilon. Thank you@ageron, that's really help.

**Comment by ageron** [comment]
My pleasure.  :) I'll close the issue, feel free to reopen if needed.
Cheers,
Aurélien
